### ISSUE TITLE
Slow mirror scans

### ISSUE THREAD

**Issue by rgaudin** [kiwix/k8s#7]
Running the `update_mirrorbrain_db.sh` script is very slow.

Some of its log is confusing to me:

```
2022-03-31T16:54:42.152549791+02:00 Scanning mirror 'dotsrc.org' at zim
2022-03-31T17:10:17.248399094+02:00 Thu Mar 31 14:55:14 2022 dotsrc.org: starting
2022-03-31T17:10:17.248434727+02:00 Thu Mar 31 14:55:16 2022 dotsrc.org: files in 'zim' before scan: 7957
2022-03-31T17:10:17.248443037+02:00 Thu Mar 31 15:10:16 2022 dotsrc.org: scanned 7957 files (8/s) in 900s
2022-03-31T17:10:17.248449878+02:00 Thu Mar 31 15:10:16 2022 dotsrc.org: files to be purged: 0
2022-03-31T17:10:17.248457167+02:00 Thu Mar 31 15:10:16 2022 dotsrc.org: total files in 'zim' after scan: 7957 (delta: 0)
2022-03-31T17:10:17.248463516+02:00 Thu Mar 31 15:10:16 2022 dotsrc.org: purged old files in 0s.
2022-03-31T17:10:17.248469968+02:00 Thu Mar 31 15:10:17 2022 dotsrc.org: done.
2022-03-31T17:10:17.252725480+02:00 Completed in 15.3 minutes
2022-03-31T17:10:18.447814836+02:00 Scanning mirror 'dotsrc.org' at zim/wikivoyage
2022-03-31T17:10:58.248454610+02:00 Thu Mar 31 15:10:50 2022 dotsrc.org: starting
2022-03-31T17:10:58.248491578+02:00 Thu Mar 31 15:10:55 2022 dotsrc.org: files in 'zim/wikivoyage' before scan: 90
2022-03-31T17:10:58.248538672+02:00 Thu Mar 31 15:10:58 2022 dotsrc.org: scanned 90 files (33/s) in 2s
2022-03-31T17:10:58.248548227+02:00 Thu Mar 31 15:10:58 2022 dotsrc.org: files to be purged: 0
2022-03-31T17:10:58.248554563+02:00 Thu Mar 31 15:10:58 2022 dotsrc.org: total files in 'zim/wikivoyage' after scan: 90 (delta: 0)
2022-03-31T17:10:58.248560832+02:00 Thu Mar 31 15:10:58 2022 dotsrc.org: purged old files in 0s.
2022-03-31T17:10:58.248566572+02:00 Thu Mar 31 15:10:58 2022 dotsrc.org: done.
2022-03-31T17:10:58.448330017+02:00 Completed in 25 seconds
```

- The first issue is the time: 15mn to scan a single mirror's `zim/` folder? 
- The second is it looks like it's doing the discovery job twice: once for all files in `zim/` then individually for each sub folder. Fortunately our tree is mostly flat so it would be done only twice for most but still.
- Third ; although it probably doesn't have much impact on the script's speed, there are many redirection errors on `ftp.acc.umu.se`:

```
2022-03-31T16:35:18.965149845+02:00 Thu Mar 31 14:32:30 2022 ftp.acc.umu.se: Error 302 occured
```

This happens for all videos for instance. It seems it is redirecting to another mirror of them but the listing is still on that mirror…

```sh
curl -I https://ftp.acc.umu.se/mirror/kiwix.org/zim/videos/aimhi_en_english-website_2022-01.zim
HTTP/1.1 302 Found
Date: Thu, 31 Mar 2022 15:58:24 GMT
Server: Apache/2.4.51 (Unix)
Location: https://laotzu.ftp.acc.umu.se/mirror/kiwix.org/zim/videos/aimhi_en_english-website_2022-01.zim
```

Might be better to register that other server as an independant (videos-only?) mirror…

**Comment by kelson42** [comment]
I confirm we have here something to do, but AFAIK we face here a limitation of mirrorbrain.

**Comment by kelson42** [comment]
Many things here:
* We should look the sync as far as possible using rsync/ftp
* We should look how to make things in paralell instead of sequencially
* It looks like the scan script do things twice as well
* ftp.acc.umu.se has a general problem linked to HTTP 302, but look at https://github.com/poeml/mirrorbrain/issues/174

Implementing all of this should allow to improve a lot the overall speed of this process.

**Comment by rgaudin** [comment]
>     * We should look the sync as far as possible using rsync/ftp

We opened a ticket at the Foundation about that. They are the only ones not providing it. Not response AFAICT

>     * We should look how to make things in paralell instead of sequencially

We've looked into it in the past and it lead to problems but we did not document it so we should try again at least to document the issue.

>     * It looks like the scan script do things twice as well

What things? Maybe opne a ticket

>     * ftp.acc.umu.se has a general problem linked to HTTP 302, but look at [Scanning an HTTPS mirror doesn't handle redirect poeml/mirrorbrain#174](https://github.com/poeml/mirrorbrain/issues/174)

There's a patch we could apply.

**Comment by kelson42** [comment]
> > ```
> > * We should look the sync as far as possible using rsync/ftp
> > ```
> 
> We opened a ticket at the Foundation about that. They are the only ones not providing it. Not response AFAICT

Yes! I guess this is https://phabricator.wikimedia.org/T348226
 
> > ```
> > * We should look how to make things in paralell instead of sequencially
> > ```
> 
> We've looked into it in the past and it lead to problems but we did not document it so we should try again at least to document the issue.

Agree, @benoit74 was talking about DB lock, but this is fearly new and unclear to me.
 
> > ```
> > * It looks like the scan script do things twice as well
> > ```
> 
> What things? Maybe opne a ticket

@rgaudin I was refering to your sentence "The second is it looks like it's doing the discovery job twice:"

> > ```
> > * ftp.acc.umu.se has a general problem linked to HTTP 302, but look at [Scanning an HTTPS mirror doesn't handle redirect poeml/mirrorbrain#174](https://github.com/poeml/mirrorbrain/issues/174)
> > ```
> 
> There's a patch we could apply.

Yes.

I would really like to be able to properly propose PR for our fixes upstream. This can only be done via a fork. In addition I guess that simple patches are OK if we have less than a handful of patches... which could come fearly quickly. At the end I don't care too much, you should know how you would like to proceed. My side I see no problem to have a `kiwix/mirrorbrain` repo.

**Comment by kelson42** [comment]
Before doing anything here, I would recommend to fix https://github.com/kiwix/container-images/issues/244 first.

**Comment by benoit74** [comment]
A first part will probably be solved by https://github.com/kiwix/container-images/issues/247

**Comment by benoit74** [comment]
@rgaudin do you recall what exactly prevented you from running scans in parallel? 

I mentioned DB lock because this is what I found somewhere while looking for advises regarding Mirrorbits and Mirrorcache, but I don't find it anymore + looking at the code it is not obvious what the issue might be, the system is at least meant to be capable to run jobs in parallel

**Comment by rgaudin** [comment]
I don't recall exactly but what I had in mind was DB locks.

**Comment by benoit74** [comment]
Some feedbacks about modifications done at the beginning of the week:
- Job duration has dropped significantly to around 40 minutes : https://kiwixorg.grafana.net/d/bb0f0990-04c5-4314-8afc-6185ac49c668/mirrorbrain?orgId=1&from=1697092766769&to=1697697566769
- Parallelization of mirrors is not yet in placed, but worked fine during a manual test (I launched a manual job when doing the modification while an "old" one was still running, the new job completed before the old one and did not got stuck)

Do you agree that we should investigate parallelization of mirrors? (this would allow to scale almost without issues, instead of today situationwhere each mirror added increase the job duration)

**Comment by benoit74** [comment]
A screenshot for archiving + lazzy ones
![image](https://github.com/kiwix/k8s/assets/7102089/b95b9c6c-9774-407a-b195-4990144d14df)

**Comment by benoit74** [comment]
Oh and I forgot to mention, but scanning is still working well 🙄 

See for instance this zim created early today:
https://download.kiwix.org/zim/wikipedia/wikipedia_tt_all_nopic_2023-10.zim?mirrorlist

**Comment by rgaudin** [comment]
Can you add that graph to the dashboard we use for routine? A glimpse a it would help detect issues

**Comment by rgaudin** [comment]
Fixed ; see #152
